### PR TITLE
Inference memory test

### PR DESCRIPTION
**Pull request by wdykas**
# What does this PR do ?
This adds optional memory checks to inference tests. It will only check the memory is within 5% up or down if the field is in the golden values. We keep breaking the inference accidentally this hopefully will at least catch if we do something really bad

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [b520ee7](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/b520ee75fe591e34b08fee2acde7cdc2ce86666a)

**Comment by wdykas** [comment]
@NVIDIA/inference thoughts? so we at least can catch when we have  bad memory regressions sometimes.

**Comment by wdykas** [comment]
/ok to test [ca2f81a](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/ca2f81a43ad0aa4762f0788f7be1a3cb6cf27a7e)

**Comment by wdykas** [comment]
/ok to test [d1b016d](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/d1b016de447e652330b4b184b8b49cc00426e51c)

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 1]
```diff
@@ -62,6 +62,19 @@
 torch.serialization.add_safe_globals([megatron.core.rerun_state_machine.RerunState])
 torch.serialization.add_safe_globals([megatron.core.rerun_state_machine.RerunDiagnostic])
 
+def _get_global_peak_memory_stats_bytes() -> dict:
```
Comment (mathemakitten) [inline examples/inference/gpt/gpt_dynamic_inference.py]: Can we move this into `examples/inference/gpt/utils.py`? I think we will be using it elsewhere as well.

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 2]
```diff
@@ -62,6 +62,19 @@
...
+
+    Uses `torch.cuda.max_memory_allocated()` and assumes peak stats were reset
+    before the benchmark run.
+    """
+    peak_alloc = int(torch.cuda.max_memory_allocated())
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        t = torch.tensor([peak_alloc], device="cuda", dtype=torch.int64)
+        torch.distributed.all_reduce(t, op=torch.distributed.ReduceOp.MAX)
+        peak_alloc = int(t[0].item())
+    return {"mem-max-allocated-bytes": peak_alloc}
```
Comment (mathemakitten) [inline examples/inference/gpt/gpt_dynamic_inference.py]: extremely petty nit: most other dictionary keys in the inference codebase use underscores, any reason not to make this `mem_max_allocated_bytes`?

File: tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py [hunk 3]
```diff
@@ -54,6 +85,35 @@ def test_inference_pipeline(golden_values_path: str, test_values_path: str) -> N
...
+        golden = _median_as_float(output_groundtruth["mem-max-allocated-bytes"])
+        assert golden > 0, f"Golden mem-max-allocated-bytes must be > 0, got {golden}."
+
+        low = 0.9 * golden
+        high = 1.1 * golden
+
+        if sampled < low:
+            raise AssertionError(
+                f"Memory is too low for mem-max-allocated-bytes: "
+                f"expected within ±10% of {golden:.0f} bytes ({_bytes_to_gib(golden):.3f} GiB) "
```
Comment (mathemakitten) [inline tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: 10% for memory seems really high to me, since this shouldn't be a non-deterministic test in theory (compared to throughput, which is sensitive to hardware variation). Additionally, with the throughput test bounds at 10%, we find that there is "golden value creep", which would be ideal to avoid here if possible. Can we tighten this bound?

Comment (wdykas) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: yeah I can try tightening.  What do you think is a good bounds on this?

Comment (wdykas) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: also do we think this can be just 100% deterministic?

Comment (mathemakitten) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: The training test in `megatron-lm/tests/functional_tests/python_test_utils/test_pretraining_regular_pipeline.py` sets these thresholds so the bounds are almost deterministic but not quite (also I see here that it uses dashes for the dict so feel free to disregard my comment on dictionary key naming above): 

```
CHECK_THRESHOLDS = {
    "iteration-time": [common.ApproximateTest(atol=0, rtol=0.05)],
    "mem-allocated-bytes": [common.ApproximateTest(atol=0, rtol=0.05)],
    "mem-max-allocated-bytes": [common.ApproximateTest(atol=0, rtol=0.05)],
    ...
    "logprobs": [common.DeterministicTest(), common.ApproximateTest(atol=0, rtol=0.05)],
}
```

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 2 above)
Comment (wdykas) [reply examples/inference/gpt/gpt_dynamic_inference.py]: done!

File: tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py (hunk 3 above)
Comment (wdykas) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: oh lol I changed to the underscores already. Should I change it back lol? 


The atol here is 0 does that mean it is deterministic?

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 1 above)
Comment (wdykas) [reply examples/inference/gpt/gpt_dynamic_inference.py]: done

File: tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py (hunk 3 above)
Comment (mathemakitten) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: This seems to suggest that a 5% difference is acceptable for training so maybe we just match this in entirety? Including the dashes in the naming, sorry 😅 We'll still have to be careful about golden value creep going forward.

Comment (wdykas) [reply tests/functional_tests/python_test_utils/test_inference_regular_pipeline.py]: ok I made the 5% change

**Comment by wdykas** [comment]
/ok to test [e3b352e](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/e3b352e7450508dff2f0f5bae1eb6ec69669529c)

**Comment by wdykas** [comment]
/ok to test [4c17ce2](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/4c17ce2239a3b3ebef10bc249f97a77748567ccf)

**Review by santhnm2 — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [f1d856f](https://github.com/NVIDIA/Megatron-LM/pull/2724/commits/f1d856ff621c267273f5ab33b9ac68ddd67a1457)

**Review by ko3n1g — APPROVED** [review]
